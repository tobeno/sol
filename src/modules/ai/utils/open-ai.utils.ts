import type {
  CreateChatCompletionRequest,
  CreateChatCompletionResponse,
  OpenAIApi,
} from 'openai';

function getOpenAi() {
  return require('openai') as typeof import('openai');
}

export function isOpenAiApiAvailable(): boolean {
  return !!process.env.OPENAI_API_KEY;
}

export function getOpenAiApi(): OpenAIApi {
  if (!isOpenAiApiAvailable()) {
    throw new Error(
      'To use the OpenAI API the OPENAI_API_KEY needs to be configured.',
    );
  }

  const openai = getOpenAi();

  return new openai.OpenAIApi(
    new openai.Configuration({
      apiKey: process.env.OPENAI_API_KEY,
    }),
  );
}

export async function createOpenAiChatCompletion(
  request: Omit<CreateChatCompletionRequest, 'model'>,
): Promise<CreateChatCompletionResponse> {
  return (
    await getOpenAiApi().createChatCompletion({
      model: 'gpt-3.5-turbo',
      ...request,
    })
  ).data;
}
