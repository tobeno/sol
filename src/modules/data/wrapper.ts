/**
 * Base class for all object based wrappers
 */
export abstract class Wrapper<ValueType = any> {
  constructor(public value: ValueType) {}

  as(name: string): this {
    (global as any)[name] = this;

    return this;
  }
}
