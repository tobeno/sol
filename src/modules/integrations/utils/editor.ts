import { File } from '../../../wrappers/file';
import { Text } from '../../../wrappers/text';
import { TmpFile } from '../../../wrappers/tmp-file';

/**
 * Opens the given path or value in the default editor.
 */
export function edit(pathOrValue?: any): File {
  let f: File;

  if (!pathOrValue) {
    f = TmpFile.create('ts');
  } else if (
    typeof pathOrValue === 'string' &&
    /(^\s|\n|\s$)/.test(pathOrValue)
  ) {
    f = TmpFile.create('txt');
    f.text = pathOrValue;
  } else if (pathOrValue instanceof Buffer) {
    f = TmpFile.create();
    f.buffer = pathOrValue;
  } else if (pathOrValue instanceof Text) {
    f = TmpFile.create(pathOrValue.ext);
    f.text = pathOrValue;
  } else if (typeof pathOrValue === 'object') {
    let ext = 'json';
    if (pathOrValue.constructor !== Object) {
      if (typeof pathOrValue.ext !== 'undefined') {
        ext = pathOrValue.ext;
      }

      if (typeof pathOrValue.text !== 'undefined') {
        pathOrValue = pathOrValue.text;
      } else if (typeof pathOrValue.data !== 'undefined') {
        pathOrValue = pathOrValue.data;
      }
    }

    f = TmpFile.create(ext);
    if (typeof pathOrValue === 'string') {
      f.text = pathOrValue;
    } else {
      f.json = pathOrValue;
    }
  } else {
    f = File.create(pathOrValue);
  }

  f.create();

  f.edit();

  return f;
}
