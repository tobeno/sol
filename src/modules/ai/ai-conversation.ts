import {
  ChatCompletionRequestMessage,
  CreateChatCompletionResponse,
} from 'openai';
import { inspect } from 'util';
import { createOpenAiChatCompletion, isOpenAiApiAvailable } from './open-ai';
import { createHash } from 'crypto';
import { Data } from '../data/data';
import { Text } from '../data/text';
import { open } from '../integrations/open';
import { getClipboard } from '../clipboard/clipboard';

export class AiConversation {
  private readonly messagesInternal: ChatCompletionRequestMessage[] = [];

  usage = Data.create({
    promptTokens: 0,
    completionTokens: 0,
    totalTokens: 0,
  });

  constructor(
    private options: {
      messages?: ChatCompletionRequestMessage[];
      dryRun?: boolean;
    } = {},
  ) {
    let { messages = [] } = options;
    if (!messages.some((message) => message.role === 'system')) {
      messages = [
        {
          role: 'system',
          content: 'You are a helpful assistant.',
        },
        ...messages,
      ];
    }

    this.messagesInternal = messages;
  }

  /**
   * Returns a MD5 has of the conversation.
   */
  private get hash(): string {
    const content = JSON.stringify(this.messagesInternal);

    return createHash('md5').update(content).digest('hex');
  }

  get messages(): Data<ChatCompletionRequestMessage[]> {
    return Data.create(this.messagesInternal);
  }

  get answers(): Data<Text[]> {
    return Data.create(
      this.messagesInternal
        .filter((message) => message.role === 'assistant')
        .map((message) => Text.create(message.content)),
    );
  }

  get answer(): Text | null {
    return (this.answers.get(-1) as Text) || null;
  }

  get answerCode(): Text | null {
    return this.answer?.selectCode() || null;
  }

  get questions(): Data<Text[]> {
    return Data.create(
      this.messagesInternal
        .filter((message) => message.role === 'user')
        .map((message) => Text.create(message.content)),
    );
  }

  get question(): Text | null {
    return (this.questions.get(-1) as Text) || null;
  }

  get questionCode(): Text | null {
    return this.question?.selectCode() || null;
  }

  get questionAndAnswer(): Text | null {
    return Text.create(
      (
        (this.question?.value || null) +
        '\n' +
        (this.answer?.value || null)
      ).trim(),
    );
  }

  get questionAndAnswerCode(): Text | null {
    return this.questionAndAnswer?.selectCode() || null;
  }

  ask(question: string): this {
    const { dryRun } = this.options;
    if (!isOpenAiApiAvailable() && !dryRun) {
      getClipboard().text = question;
      open('http://chat.openai.com');
      return this;
    }

    this.messagesInternal.push({
      role: 'user',
      content: question,
    });

    let answer: string | null = null;
    if (!dryRun) {
      const cacheFile = solUserWorkspace.cacheDir.file(
        `ai-conversation-response-${this.hash}.json`,
      );

      let response: CreateChatCompletionResponse;
      if (!cacheFile.exists) {
        response = createOpenAiChatCompletion({
          messages: this.messagesInternal,
        });

        cacheFile.json = response;
      } else {
        response = cacheFile.json.value;
      }

      answer = response.choices[0].message?.content || null;
      if (!answer) {
        throw new Error(`No answer received.

Response: ${JSON.stringify(response, null, 2)}`);
      }

      const usage = this.usage.value;
      usage.promptTokens += response.usage?.prompt_tokens || 0;
      usage.completionTokens += response.usage?.completion_tokens || 0;
      usage.totalTokens += response.usage?.total_tokens || 0;
    } else {
      answer = 'This is a dry run';
    }

    this.messagesInternal.push({
      role: 'assistant',
      content: answer,
    });

    return this;
  }

  askCode(question: string): this {
    return this.ask(`${question}\nONLY CODE, NOTHING ELSE`);
  }

  [inspect.custom](): string {
    return this.toString();
  }

  toString(): string {
    return this.messagesInternal
      .filter((message) => message.role !== 'system')
      .map((message) => {
        let role: string = message.role;
        if (role === 'assistant') {
          role = 'ðŸ¤–';
        } else if (role === 'user') {
          role = 'ðŸ¤”';
        } else if (role === 'system') {
          role = 'ðŸ’»';
        }

        return `${role}: ${message.content}`;
      })
      .join('\n---\n');
  }
}
