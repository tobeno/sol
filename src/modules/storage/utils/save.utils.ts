import { File } from '../../../wrappers/file.wrapper';
import { Text } from '../../../wrappers/text.wrapper';
import { TmpFile } from '../../../wrappers/tmp-file.wrapper';

/**
 * Saves the given content to a file.
 */
export function save(content: any): File {
  let file = null;
  let text: Text | null = null;
  const textGeneric = content.text;
  if (textGeneric instanceof Text) {
    text = textGeneric;
  }

  if (!text) {
    throw new Error('Could not convert to text');
  }

  const rootSource = content.rootSource;
  if (rootSource) {
    if (rootSource instanceof File) {
      file = rootSource;
    } else {
      throw new Error('Source does not support saving');
    }
  }

  if (!file) {
    file = TmpFile.create(text.ext);
  }

  file.text = text;

  return file;
}

/**
 * Saves the given content to a file with the given path.
 */
export function saveAs(content: any, path: string): File {
  const f = File.create(path);
  f.text = content;
  return f;
}
