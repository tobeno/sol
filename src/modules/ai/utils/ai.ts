import { AiConversation } from '../wrappers/ai-conversation';

export function askAi(question: string): AiConversation {
  return new AiConversation().ask(question);
}

export function askAiCode(question: string): AiConversation {
  return new AiConversation().askCode(question);
}
