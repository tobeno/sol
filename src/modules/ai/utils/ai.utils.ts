import { AiConversation } from '../wrappers/ai-conversation.wrapper';

export async function askAi(question: string): Promise<AiConversation> {
  return new AiConversation().ask(question);
}

export async function askAiCode(question: string): Promise<AiConversation> {
  return new AiConversation().askCode(question);
}
