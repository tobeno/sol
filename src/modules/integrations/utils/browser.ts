import { open } from '../../../utils/open';
import { Url } from '../../../wrappers/url';
import { Wrapper } from '../../../wrappers/wrapper';

/**
 * Opens the given URL in the default browser.
 */
export function browse(url?: any): Url {
  if (url instanceof Wrapper) {
    url = url.value;
  }

  if (!url?.startsWith('http')) {
    open(url, 'chrome');
  } else {
    open(url || 'https://www.google.com');
  }

  return Url.create(url);
}
