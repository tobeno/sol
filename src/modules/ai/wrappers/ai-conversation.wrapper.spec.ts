import { describe, expect, it } from 'vitest';
import { unwrapDeep } from '../../../utils/wrapper.utils';
import { AiConversation } from './ai-conversation.wrapper';

describe('AiConversation', () => {
  describe('ask', () => {
    it('should ask a question', async () => {
      const conversation = await new AiConversation({
        dryRun: true,
      }).ask('Hello');
      expect(conversation).toBeInstanceOf(AiConversation);
      expect(unwrapDeep(conversation.messages)).toMatchSnapshot('messages');
      expect(unwrapDeep(conversation.answers)).toMatchSnapshot('answers');
      expect(unwrapDeep(conversation.questions)).toMatchSnapshot('questions');
    });

    it('should ask a question with custom system message', async () => {
      const conversation = await new AiConversation({
        messages: [
          {
            role: 'system',
            content: 'You are a test assistant.',
          },
        ],
        dryRun: true,
      }).ask('Hello');
      expect(conversation).toBeInstanceOf(AiConversation);
      expect(unwrapDeep(conversation.messages)).toMatchSnapshot('messages');
      expect(unwrapDeep(conversation.answers)).toMatchSnapshot('answers');
      expect(unwrapDeep(conversation.questions)).toMatchSnapshot('questions');
    });
  });

  describe('askCode', () => {
    it('should ask a question', async () => {
      const conversation = await new AiConversation({
        dryRun: true,
      }).askCode('Hello');
      expect(conversation).toBeInstanceOf(AiConversation);
      expect(unwrapDeep(conversation.messages)).toMatchSnapshot('messages');
      expect(unwrapDeep(conversation.answers)).toMatchSnapshot('answers');
      expect(unwrapDeep(conversation.questions)).toMatchSnapshot('questions');
    });

    it('should ask a question with custom system message', async () => {
      const conversation = await new AiConversation({
        messages: [
          {
            role: 'system',
            content: 'You are a test assistant.',
          },
        ],
        dryRun: true,
      }).askCode('Hello');
      expect(conversation).toBeInstanceOf(AiConversation);
      expect(unwrapDeep(conversation.messages)).toMatchSnapshot('messages');
      expect(unwrapDeep(conversation.answers)).toMatchSnapshot('answers');
      expect(unwrapDeep(conversation.questions)).toMatchSnapshot('questions');
    });
  });
});
