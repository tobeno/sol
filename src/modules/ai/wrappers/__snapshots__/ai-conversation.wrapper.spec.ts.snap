// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AiConversation > ask > should ask a question > answers 1`] = `
[
  "This is a dry run",
]
`;

exports[`AiConversation > ask > should ask a question > messages 1`] = `
[
  {
    "content": "You are a helpful assistant.",
    "role": "system",
  },
  {
    "content": "Hello",
    "role": "user",
  },
  {
    "content": "This is a dry run",
    "role": "assistant",
  },
]
`;

exports[`AiConversation > ask > should ask a question > questions 1`] = `
[
  "Hello",
]
`;

exports[`AiConversation > ask > should ask a question with custom system message > answers 1`] = `
[
  "This is a dry run",
]
`;

exports[`AiConversation > ask > should ask a question with custom system message > messages 1`] = `
[
  {
    "content": "You are a test assistant.",
    "role": "system",
  },
  {
    "content": "Hello",
    "role": "user",
  },
  {
    "content": "This is a dry run",
    "role": "assistant",
  },
]
`;

exports[`AiConversation > ask > should ask a question with custom system message > questions 1`] = `
[
  "Hello",
]
`;

exports[`AiConversation > askCode > should ask a question > answers 1`] = `
[
  "This is a dry run",
]
`;

exports[`AiConversation > askCode > should ask a question > messages 1`] = `
[
  {
    "content": "You are a helpful assistant.",
    "role": "system",
  },
  {
    "content": "Hello
ONLY CODE, NOTHING ELSE",
    "role": "user",
  },
  {
    "content": "This is a dry run",
    "role": "assistant",
  },
]
`;

exports[`AiConversation > askCode > should ask a question > questions 1`] = `
[
  "Hello
ONLY CODE, NOTHING ELSE",
]
`;

exports[`AiConversation > askCode > should ask a question with custom system message > answers 1`] = `
[
  "This is a dry run",
]
`;

exports[`AiConversation > askCode > should ask a question with custom system message > messages 1`] = `
[
  {
    "content": "You are a test assistant.",
    "role": "system",
  },
  {
    "content": "Hello
ONLY CODE, NOTHING ELSE",
    "role": "user",
  },
  {
    "content": "This is a dry run",
    "role": "assistant",
  },
]
`;

exports[`AiConversation > askCode > should ask a question with custom system message > questions 1`] = `
[
  "Hello
ONLY CODE, NOTHING ELSE",
]
`;
